File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    
    int getLimit(map<int,int>&mp){
        
        return mp.rbegin()->first - mp.begin()->first;
    }
    
    int countAtMostK(vector<int>&nums,int k){

        int ans=0,start=0;
        map<int,int>mp;
        
        for(int i=0;i<nums.size();i++){
            
            mp[nums[i]]++;
            
            while(mp.size() and getLimit(mp)>k){
                int curr = nums[start];
                if(mp[curr]==1)
                    mp.erase(curr);
                
                else mp[curr]--;
                
                start++;
            }
            
            
            ans= max(ans,(i - start + 1));
        }
        
        return ans;
    }
    
    int longestSubarray(vector<int>& nums, int limit) {
        
        return countAtMostK(nums,limit);
    }
};

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        
        int n = nums.length, start = 0, ans = 0;
        TreeMap<Integer, Integer> mp = new TreeMap<>();
        
        for (int i = 0; i < n; i++) {
            mp.put(nums[i], mp.getOrDefault(nums[i], 0) + 1);
            
            while (mp.size() > 1 && (mp.lastEntry().getKey() - mp.firstEntry().getKey() > limit)) {
                int currNum = nums[start];
                
                if (mp.get(currNum) == 1) {
                    mp.remove(currNum);
                } else {
                    mp.put(currNum, mp.get(currNum) - 1);
                }
                
                start++;
            }
            
            ans = Math.max(ans, i - start + 1);
        }
        
        return ans;
    }
}
